Add per viz-type card template sizes in x-rays

markbastian opened the pull request:
Previously, any card without a size (width or height) set in their yaml definition was set to a default of `default-card-width` or `default-card-width`, respectively. This PR adds a map of default and min dimensions per viz type (the `:display`) key and fills that in for each template. The template is merged into this result, keeping anything that already existed in the yaml file.

Prior to this PR, the default card sizes were found in `sizes.ts` and were inaccessible from the BE. This PR consolidates those changes into `metabase.shared.automagic-dashboards.constants`.

Note that @kulyk and @JesseSDevaney might want to consider further consolidation of dashboard related constants into this namespace.

markbastian committed:
Add per viz-type card template sizes in x-rays

Previously, any card without a size (width or height) set in their yaml definition was set to a default of `default-card-width` or `default-card-width`, respectively. This PR adds a map of default and min dimensions per viz type (the `:display`) key and fills that in for each template. The template is merged into this result, keeping anything that already existed in the yaml file.

Before merging this we should consider if we want to use the min values and where this data should be stored. It is currently in `metabase.automagic-dashboards.dashboard-templates`, but was lifted from [sizes.ts](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/shared/utils/sizes.ts).

markbastian committed:
Moving card sizing constants to cljc

The default card sizes were found in `sizes.ts` and were duplicated into `metabase.automagic-dashboards.dashboard-templates`. This PR consolidates those changes into `metabase.shared.automagic-dashboards.constants`.

Note that @kulyk and @JesseSDevaney might want to consider further consolidation of dashboard related constants into this namespace.

bshepherdson commented on src/metabase/automagic_dashboards/dashboard_templates.clj:
```diff
@@ -286,6 +287,20 @@
+  [dashboard-template]
+  (update dashboard-template :cards
+          (fn [cards]
+            (mapv
+              (fn [card-spec]
+                (update-vals
+                  card-spec
+                  (fn [{:keys [visualization] :as card-spec}]
+                    (let [defaults (get-in magic.constants/card-size-defaults [(keyword visualization) :default])]
+                      (into defaults card-spec)))))
```
This function would be a lot easier to read if this were a named helper. Then the outer function becomes `(update dashboard-template :cards #(mapv new-helper %))`.

markbastian replied:
Done

bshepherdson commented on src/metabase/shared/automagic_dashboards/constants.cljc:
```diff
@@ -0,0 +1,34 @@
+   :heading     {:min {:width 1 :height 1} :default {:width GRID_WIDTH}}
+   :text        {:min {:width 1 :height 1} :default {:width 12 :height 3}}})
+
+(def ^:export card-size-defaults
+  "Default card sizes per visualization type"
+  #?(:clj  card-size-defaults*
+     :cljs (clj->js card-size-defaults*)))
```
I'd do this same-data-different-format magic the other way around. Rename `card-size-defaults*` to remove the star - that's the symbol for use from **all** Clojure code. Then add
```clojure
#?(:cljs (def ^:export card-size-defaults-json (clj->js card-size-defaults)))
```
and import that from TS.

I'd certainly be surprised if I were working in some CLJC file, `:require`d some CLJC namespace, and received JS data! We use the approach I'm suggesting in MLv2: most of the files are all Clojure all the time and have very few `#?(:cljs ...)` conditionals, then `metabase.lib.js` wraps many of those functions with extra JS conversion for inputs and outputs.

markbastian replied:
That make sense.

markbastian committed:
Incorporating Feedback

Changes
- Explicitly defined two constants in the shared file that are clj and js friendly
- Set case to CARD_SIZE_DEFAULTS_JSON
- Updated `metabase.automagic-dashboards.dashboard-templates/set-default-card-dimensions` to use a helper

JesseSDevaney commented on src/metabase/shared/automagic_dashboards/constants.cljc:
Nit: Can we change the namespace name to be `.dashboards.constants` and re-locate the file to `./shared/dashboards/constants.cljc` so that it is more obvious that these constants apply to all dashboards, not just X-Rays?

JesseSDevaney commented on src/metabase/shared/automagic_dashboards/constants.cljc:
```diff
@@ -0,0 +1,34 @@
+   :map         {:min {:width 4 :height 3} :default {:width 12 :height 6}}
+   :object      {:min {:width 4 :height 3} :default {:width 12 :height 9}}
+   :row         {:min {:width 4 :height 3} :default {:width 12 :height 6}}
+   :heading     {:min {:width 1 :height 1} :default {:width GRID_WIDTH}}
```
```suggestion
   :heading     {:min {:width 1 :height 1} :default {:width GRID_WIDTH :height 1}}
```

JesseSDevaney reviewed (changes requested):
99.99% LGTM

There is one thing we should address before merging, and a small nit.

markbastian committed:
Update src/metabase/shared/automagic_dashboards/constants.cljc

Co-authored-by: Jesse Devaney <22608765+JesseSDevaney@users.noreply.github.com>

markbastian committed:
Updating ns name per Jesse
